src/Remote/Importer.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+namespace Castor\Remote;
+
+use Castor\GlobalHelper;
+use Castor\PathHelper;
+use Castor\Remote\Exception\ComposerError;
+use Castor\Remote\Exception\ImportError;
+use Castor\Remote\Exception\InvalidImportFormat;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Process\ExecutableFinder;
+
+/** @internal */
+class Importer
+{
+    public function __construct(
+        private readonly LoggerInterface $logger,
+        private readonly Composer $composer,
+        private readonly Filesystem $filesystem,
+    ) {
+    }
+
+    /**
+     * @param ?array{
+     *     url?: string,
+     *     type?: "git" | "svn",
+     *     reference?: string,
+     * } $source
+     */
+    public function importFunctionsFrom(string $scheme, string $package, ?string $version = null, ?string $vcs = null, ?array $source = null): string
+    {
+        if (!$this->allowsRemote()) {
+            throw new ImportError(sprintf('Remote imports are disabled, skipping import of "%s".', $package));
+        }
+
+        $composer = (new ExecutableFinder())->find('composer');
+
+        if (!$composer) {
+            throw new ImportError('The "composer" executable was not found. Please make sure that Composer is installed and available in your PATH.');
+        }
+
+        if (!preg_match('#^(?<organization>[^/]+)/(?<repository>[^/]+)$#', $package)) {
+            throw new InvalidImportFormat(sprintf('The import path must be formatted like this: "%s://<organization>/<repository>".', $scheme));
+        }
+
+        if ('composer' === $scheme) {
+            if (null !== $source) {
+                throw new InvalidImportFormat('The "source" argument is not supported for Composer/Packagist packages.');
+            }
+
+            return $this->importPackageWithComposer($package, version: $version ?? '*', repositoryUrl: $vcs);
+        }
+        if ('package' === $scheme) {
+            if (null !== $version || null !== $vcs) {
+                throw new InvalidImportFormat('The "source" and "vcs" arguments are not supported for non-Composer packages.');
+            }
+            if (null === $source) {
+                throw new InvalidImportFormat('The "source" argument is required for non-Composer packages.');
+            }
+
+            return $this->importPackageWithComposer($package, version: 'v1', source: $source);
+        }
+
+        throw new InvalidImportFormat(sprintf('The import scheme "%s" is not supported.', $scheme));
+    }
+
+    public function reset(): void
+    {
+        $this->composer->setConfiguration([
+            'description' => 'This file is managed by Castor. Do not edit it manually.',
+            'config' => [
+                'sort-packages' => true,
+            ],
```

Comment:
We should add this : 

```
'version' => CURRENT_CASTOR_VERSION,
'replace' => [
    "castor/castor" => "self.version",
],
```

So we do not need to download castor if an extension require it, and it allow extension developer to add constraint to the castor version which will fail if the current castor version is not compatible

Reply:
We could use put the version directly in the replace, so the root version can be something else that will be used in the future when someone want to compile their own castor ?

Reply:
Done :+1: